libraries/chain/asset_evaluator.cpp

```diff
@@ -288,6 +289,18 @@ void_result asset_update_evaluator::do_evaluate(const asset_update_operation& o)
    FC_ASSERT(!((o.new_options.flags ^ a.options.flags) & ~a.options.issuer_permissions),
              "Flag change is forbidden by issuer permissions");
 
+   // TODO HARDFORKCHECK
+   // change of the disable_modify_max_supply flag; should only be set from 0 to 1
+   FC_ASSERT( ( old_options.flags & disable_modify_max_supply == o.new_options.flags & disable_modify_max_supply
+           || ( old_options.flags & disable_modify_max_supply ) == 0 ),
+           "The disable_modify_max_supply flag can not be deactivated." );
```

Comment:
@dimfred: could you elaborate what this is about?

Reply:
So with our flag we want to block the issuer's ability to modify the `max_supply` of his created asset. However it is optional for the issuer to set the flag. But once set, the flag can't be reset, to guarantee that the promise (to not modify the `max_supply`) is kept. 

So with the first statement `old_options.flags & disable_modify_max_supply == o.new_options.flags & disable_modify_max_supply`, the flag in the current state of the `asset.options` (`old_options`) is compared to the flag in the new state of the `asset.options` (`o.new_options`).  This means the flag isn't modified and thus results in a valid option setup.

The second case `old_options.flags & disable_modify_max_supply ) == 0` is true when the flag is 0. When it's 1 (modification disabled) and should be reset ( first statement == false ) the statement will be false, because like said, it is permitted to reset the flag.

Reply:
I'd recommend use issuer_permission for this kind of rule.

Reply:
What @abitmore says!

As long as the permission is set (default), the flag can be toggeled at any time.
Once the permission is removed, the flag is frozen indefinitely.

That behavior is already define here:
https://github.com/bitshares/bitshares-core/pull/1375/files/4d50ac6b29875b81b8fed70a135ffc69c444895b#diff-6b07a69176a9fc1a8c6ea1087e4129adR289

Reply:
Agree.

I would prefer to invert the meaning of the flag though, i. e. call it allow_modify_max_supply. Negative logic is always confusing.

In addition, the permission (and flag, if inverted as suggested) must be enabled for all existing assets.

Reply:
@pmconrad That was our first thought too, but every other `issuer_permissions` are also defined as disabled. Also if you for example create a new asset and forget to set the flag, the asset will be created with fixed supply by default.  But in this case there won't be any `current_supply` if understand it right. So with the above mentioned logic the issuer would still be able to modify  the `max_supply` but not be able to reset the flag (or in the `allow_modify_max_supply`case to enable it. Maybe it  would also make sense to be able to reset the `disable_modify_max_supply` , when there is no `current_supply`.

Reply:
AFAICS only two out of 9 flags have negative logic: https://github.com/bitshares/bitshares-core/blob/2.0.180823/libraries/chain/include/graphene/chain/protocol/types.hpp#L96-L104

Reply:
Yeah you are right, but if we switch the flag to be a non inverted logic, it would be not activated by default. AFAICS all flags are 0 by default. This means for us if we make the flag `enable_modify_max_supply` the modification would be disabled by default. This would change the current behaviour of the `asset_create_operation`. 

I think it would be possible to set the flag in the `asset_evaluator::do_apply` to make it default on creation, but then there would be more code and it would break the current pattern of having all flags 0 by default. But if you say this is okay, I agree with you.

Reply:
My assumption is that the newly introduced flat is initialized with `0` for existing assets.

To keep current behavior for existing assets, we *have to* use a `disable_*` flag which changes from the current behavior when set.

Else, if we were to add an `enable_issue_shares` operation, and hard forked over, daily operations of CryptoBridge/OpenLedger/GDEX etc would break, not being able to issue new shares after hard fork.

Hence, the `disable_*` methodology.

Reply:
We can automatically set the permission + flag on asset_create and asset_update until the hardfork takes effect. Slower alternative would be to switch them on at the time of the hardfork.

IMO the little extra code is well worth the clarity we gain.

There isn't really a default for the flags/permissions, the value *must* be specified at asset creation.

Reply:
@pmconrad again please be aware that the client libraries and all the 3rd-party software that depends on the libraries would need to adapt to the change. If we use the `disable_` methodology, unchanged clients will remain old behavior when creating new assets. This may save efforts greatly and avoid potential troubles.

Reply:
Good point.